training/deepspeech_training/util/feeding.py

```diff
@@ -99,15 +101,19 @@ def generate_values():
         epoch = epoch_counter['epoch']
         if train_phase:
             epoch_counter['epoch'] += 1
-        samples = samples_from_sources(sources, buffering=buffering, labeled=True)
+        samples = samples_from_sources(sources, buffering=buffering, labeled=True, reverse=reverse)
         num_samples = len(samples)
+        if limit > 0:
+            num_samples = min(limit, num_samples)
         samples = apply_sample_augmentations(samples,
                                              augmentations,
                                              buffering=buffering,
                                              process_ahead=2 * batch_size if process_ahead is None else process_ahead,
                                              clock=epoch / epochs,
                                              final_clock=(epoch + 1) / epochs)
         for sample_index, sample in enumerate(samples):
+            if sample_index >= limit > 0:
```

Comment:
Why not just use `sample_index >= num_samples` here? The way Python works makes this unintuitive to read: "if `limit` is in the interval `[sample_index, 0)`... but wait, `limit` is constant, it's `sample_index` that's changing...".

```suggestion
            if sample_index >= num_samples:
```